File: packages/crypto/scripts/gen-vectors.mjs
/**
 * Writes the TypeScript half of the cross-language test vectors that hold
 * this package and the Rust crate `engram-core` to the same bytes.
 *
 * Two kinds of artifact, in two files:
 *  - vectors.json: deterministic outputs for fixed inputs. Regenerated on
 *    every run; CI regenerates and fails if the committed copy moved, so
 *    no format change can slip through unnoticed.
 *  - sealed.json (+ .bin sidecars): artifacts whose API draws randomness
 *    (nonces, stream headers, ephemeral keys). Written once and kept; they
 *    are inputs for the Rust side to open, not values to reproduce.
 *
 * Run from packages/crypto:  node scripts/gen-vectors.mjs
 */
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  ready,
  toB64,
  fromB64,
  contentDigest,
  createDigester,
  deriveKeyEncryptionKey,
  deriveLoginKey,
  loginKeyDigest,
  deriveUnlockKey,
  secretBoxSeal,
  generateKeyPair,
  sealToPublicKey,
  encryptBytes,
  chunkedEncrypt,
  chunkedCiphertextSize,
  streamCiphertextSize,
  streamPlaintextSize,
  encryptFileMetadata,
} from "../src/index.ts";

await ready();

const here = dirname(fileURLToPath(import.meta.url));
const tsDir = join(here, "..", "test", "vectors", "ts");
await mkdir(tsDir, { recursive: true });

const pattern = (n, mul, add) => {
  const out = new Uint8Array(n);
  for (let i = 0; i < n; i++) {
    out[i] = (i * mul + add) % 256;
  }
  return out;
};

const KEY = pattern(32, 13, 5);
const SECRET = pattern(48, 7, 3);
const SALT = pattern(16, 11, 2);

// ---- deterministic ----

const kdfFloor = { salt: toB64(SALT), opsLimit: 2, memLimit: 19 * 1024 * 1024 };
const kdfMid = { salt: toB64(SALT), opsLimit: 3, memLimit: 64 * 1024 * 1024 };
const kekFloor = deriveKeyEncryptionKey("correct horse battery staple", kdfFloor).kek;
const kekMid = deriveKeyEncryptionKey("pässwörd — unicode", kdfMid).kek;
const loginKey = deriveLoginKey(kekFloor);

const vectors = {
  b64url: [
    { hexIn: "", out: "" },
    { hexIn: "00", out: toB64(fromHexStr("00")) },
    { hexIn: "0001", out: toB64(fromHexStr("0001")) },
    { hexIn: "fbff7e", out: toB64(fromHexStr("fbff7e")) },
    { hexIn: bytesToHex(KEY), out: toB64(KEY) },
  ],
  digest: [
    { hexIn: "", out: contentDigest(new Uint8Array(0)) },
    { hexIn: bytesToHex(pattern(100, 3, 1)), out: contentDigest(pattern(100, 3, 1)) },
    {
      incremental: [bytesToHex(pattern(10, 1, 0)), bytesToHex(pattern(20, 5, 9))],
      out: (() => {
        const d = createDigester();
        d.update(pattern(10, 1, 0));
        d.update(pattern(20, 5, 9));
        return d.final();
      })(),
    },
  ],
  argon2id: [
    { password: "correct horse battery staple", ...kdfFloor, kek: toB64(kekFloor) },
    { password: "pässwörd — unicode", ...kdfMid, kek: toB64(kekMid) },
  ],
  kdf: {
    kek: toB64(kekFloor),
    loginKey,
    loginKeyDigest: loginKeyDigest(loginKey),
    unlockSecretHex: bytesToHex(SECRET),
    unlockKey: toB64(deriveUnlockKey(SECRET)),
  },
  sizes: [0, 1, 100, 4 * 1024 * 1024, 4 * 1024 * 1024 + 1, 9_000_000].map((n) => ({
    plain: n,
    chunked: chunkedCiphertextSize(n),
    stream: streamCiphertextSize(n),
    streamBack: streamPlaintextSize(streamCiphertextSize(n)),
  })),
};

await writeFile(join(tsDir, "vectors.json"), JSON.stringify(vectors, null, 1));
console.log("vectors.json written");

// ---- sealed (write once) ----

const sealedPath = join(tsDir, "sealed.json");
if (existsSync(sealedPath)) {
  console.log("sealed.json already present, kept as committed");
} else {
  const pair = generateKeyPair();
  const streamPlain = pattern(100_000, 17, 11);
  const streamBlob = encryptBytes(streamPlain, KEY);
  const egcPlain = pattern(4 * 1024 * 1024 + 3, 31, 7);
  const egcBlob = chunkedEncrypt(egcPlain, KEY);
  const meta = {
    name: "fixture née photo.heic",
    mime: "image/heic",
    size: 12345,
    mtime: 1_754_500_000_000,
    tags: ["album:fixtures", "sunny"],
    favorite: true,
    digest: contentDigest(streamPlain),
    futureField: { keep: "me" },
  };
  const sealed = {
    keyHex: bytesToHex(KEY),
    secretbox: { plainHex: bytesToHex(pattern(64, 9, 4)), box: secretBoxSeal(pattern(64, 9, 4), KEY) },
    stream: { plainMul: 17, plainAdd: 11, plainLen: streamPlain.length, file: "stream.bin" },
    egc1: { plainMul: 31, plainAdd: 7, plainLen: egcPlain.length, file: "egc1.bin" },
    sealedbox: {
      publicKey: pair.publicKey,
      privateKey: toB64(pair.privateKey),
      message: "sealed across languages",
      sealed: sealToPublicKey(new TextEncoder().encode("sealed across languages"), pair.publicKey),
    },
    metadata: { value: meta, box: encryptFileMetadata(meta, KEY) },
  };
  await writeFile(sealedPath, JSON.stringify(sealed, null, 1));
  await writeFile(join(tsDir, "stream.bin"), streamBlob);
  await writeFile(join(tsDir, "egc1.bin"), egcBlob);
  console.log("sealed.json + blobs written");
}

function bytesToHex(bytes) {
  return [...bytes].map((b) => b.toString(16).padStart(2, "0")).join("");
}
function fromHexStr(hex) {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

File: crates/engram-core/src/bin/gen-vectors.rs
//! Writes the Rust half of the cross-language vectors; the mirror of
//! `packages/crypto/scripts/gen-vectors.mjs`, same two-file split:
//! deterministic `vectors.json` (regenerated, diffed in CI) and
//! `sealed.json` plus binary sidecars (written once, opened by the
//! TypeScript side).

use engram_core::{b64, chunked, digest, keys, metadata, secretbox, sealedbox, stream};
use serde_json::{json, Value};
use std::fs;
use std::path::PathBuf;

fn pattern(n: usize, mul: usize, add: usize) -> Vec<u8> {
    (0..n).map(|i| ((i * mul + add) % 256) as u8).collect()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

fn sizes() -> Vec<Value> {
    [0u64, 1, 100, 4 * 1024 * 1024, 4 * 1024 * 1024 + 1, 9_000_000]
        .iter()
        .map(|n| {
            json!({ "plain": n, "chunked": chunked::ciphertext_size(*n),
                     "stream": stream::ciphertext_size(*n) })
        })
        .collect()
}

fn main() {
    engram_core::init();
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../packages/crypto/test/vectors/rs");
    fs::create_dir_all(&dir).expect("vectors dir");

    let key: [u8; 32] = pattern(32, 13, 5).try_into().unwrap();
    let salt16: [u8; 16] = pattern(16, 11, 2).try_into().unwrap();
    let secret = pattern(48, 7, 3);
    let nonce: [u8; 24] = pattern(24, 19, 6).try_into().unwrap();

    // ---- deterministic ----
    let kek_floor = keys::derive_kek("correct horse battery staple", &salt16, 2, 19 * 1024 * 1024)
        .expect("argon2id floor");
    let kek_mid =
        keys::derive_kek("pässwörd — unicode", &salt16, 3, 64 * 1024 * 1024).expect("argon2id mid");
    let login_key = keys::derive_login_key(&kek_floor);

    let egc_plain = pattern(4 * 1024 * 1024 + 3, 29, 3);
    let egc_blob = chunked::encrypt(&egc_plain, &key, &salt16);

    let incremental = {
        let mut d = digest::Digester::new();
        d.update(&pattern(10, 1, 0));
        d.update(&pattern(20, 5, 9));
        d.finish()
    };

    let vectors = json!({
        "b64url": [
            { "hexIn": "", "out": b64::to_b64url(&[]) },
            { "hexIn": "00", "out": b64::to_b64url(&[0]) },
            { "hexIn": "0001", "out": b64::to_b64url(&[0, 1]) },
            { "hexIn": "fbff7e", "out": b64::to_b64url(&[0xfb, 0xff, 0x7e]) },
            { "hexIn": hex(&key), "out": b64::to_b64url(&key) },
        ],
        "b64std": { "hexIn": "fbff7e00", "out": b64::to_b64std(&[0xfb, 0xff, 0x7e, 0]) },
        "digest": [
            { "hexIn": "", "out": digest::digest(&[]) },
            { "hexIn": hex(&pattern(100, 3, 1)), "out": digest::digest(&pattern(100, 3, 1)) },
            { "incremental": [hex(&pattern(10, 1, 0)), hex(&pattern(20, 5, 9))], "out": incremental },
        ],
        "argon2id": [
            { "password": "correct horse battery staple", "salt": b64::to_b64url(&salt16),
              "opsLimit": 2, "memLimit": 19 * 1024 * 1024, "kek": b64::to_b64url(&kek_floor) },
            { "password": "pässwörd — unicode", "salt": b64::to_b64url(&salt16),
              "opsLimit": 3, "memLimit": 64 * 1024 * 1024, "kek": b64::to_b64url(&kek_mid) },
        ],
        "kdf": {
            "kek": b64::to_b64url(&kek_floor),
            "loginKey": b64::to_b64url(&login_key),
            "loginKeyDigest": b64::to_b64url(&keys::login_key_digest(&login_key)),
            "unlockSecretHex": hex(&secret),
            "unlockKey": b64::to_b64url(&keys::derive_unlock_key(&secret)),
            "shareSubkey1": b64::to_b64url(&keys::share_subkey(&kek_floor, 1)),
            "shareSubkey2": b64::to_b64url(&keys::share_subkey(&kek_floor, 2)),
        },
        "secretbox": {
            "keyHex": hex(&key), "nonceHex": hex(&nonce), "plainHex": hex(&pattern(64, 9, 4)),
            "box": secretbox::seal_with_nonce(&pattern(64, 9, 4), &key, &nonce),
        },
        "egc1": { "keyHex": hex(&key), "saltHex": hex(&salt16), "plainMul": 29, "plainAdd": 3,
                   "plainLen": egc_plain.len(), "file": "egc1.bin" },
        "sizes": sizes(),
    });
    fs::write(dir.join("vectors.json"), serde_json::to_string_pretty(&vectors).unwrap())
        .expect("write vectors.json");
    fs::write(dir.join("egc1.bin"), &egc_blob).expect("write egc1.bin");
    println!("vectors.json + egc1.bin written");

    // ---- sealed (write once) ----
    let sealed_path = dir.join("sealed.json");
    if sealed_path.exists() {
        println!("sealed.json already present, kept as committed");
        return;
    }
    // The TypeScript side committed a keypair; seal to its public key so
    // its consumer test can open with the matching secret.
    let ts_sealed: Value = serde_json::from_str(
        &fs::read_to_string(dir.parent().unwrap().join("ts/sealed.json"))
            .expect("ts/sealed.json exists (run the TypeScript generator first)"),
    )
    .unwrap();
    let ts_public = ts_sealed["sealedbox"]["publicKey"].as_str().unwrap();

    let stream_plain = pattern(100_000, 23, 15);
    let stream_blob = stream::encrypt_bytes(&stream_plain, &key);

    let meta = metadata::FileMetadata {
        name: "rust née fixture.pdf".into(),
        mime: "application/pdf".into(),
        size: 4242,
        mtime: 1_754_600_000_000,
        tags: Some(vec!["album:fixtures".into(), "rusty".into()]),
        favorite: Some(false),
        digest: Some(digest::digest(&stream_plain)),
        source_id: Some("asset-0001".into()),
        ..Default::default()
    };

    let sealed = json!({
        "keyHex": hex(&key),
        "stream": { "plainMul": 23, "plainAdd": 15, "plainLen": stream_plain.len(), "file": "stream.bin" },
        "sealedbox": {
            "toPublicKey": ts_public,
            "message": "sealed from rust",
            "sealed": sealedbox::seal_to_public_key(b"sealed from rust", ts_public).unwrap(),
        },
        "metadata": { "value": meta, "box": metadata::encrypt_file_metadata(&meta, &key) },
    });
    fs::write(sealed_path, serde_json::to_string_pretty(&sealed).unwrap()).expect("write sealed.json");
    fs::write(dir.join("stream.bin"), &stream_blob).expect("write stream.bin");
    println!("sealed.json + stream.bin written");
}
